File: daily-topic/increasingBST.go
package daily_topic

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func increasingBST1(root *TreeNode) *TreeNode {
	newRoot := &TreeNode{}
	r := newRoot
	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		r.Right = node
		r.Left = nil
		r = node
		inorder(node.Right)
	}
	inorder(root)
	return newRoot.Right
}

func increasingBST2(root *TreeNode) *TreeNode {
	dummyNode := &TreeNode{}
	resNode := dummyNode

	var inorder func(*TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)

		// 在中序遍历的过程中修改节点指向
		resNode.Right = node
		node.Left = nil
		resNode = node

		inorder(node.Right)
	}
	inorder(root)
	return dummyNode.Right
}

File: algo-java/src/专项突破/树/二叉搜索树/IncreasingBST.java
package 专项突破.树.二叉搜索树;

import struct.TreeNode.TreeNode;

import java.util.Queue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * @author Zexho
 * @date 2021/12/23 8:12 PM
 */
public class IncreasingBST {
    Queue<TreeNode> queue = new LinkedBlockingQueue<>();

    public TreeNode increasingBST(TreeNode root) {
        dfs(root);
        TreeNode h = queue.peek();
        while (true) {
            TreeNode poll = queue.poll();
            poll.left = null;
            if (!queue.isEmpty()) {
                poll.right = queue.peek();
            } else {
                break;
            }
        }
        return h;
    }

    private void dfs(TreeNode node) {
        if (node == null) return;
        dfs(node.left);
        queue.offer(node);
        dfs(node.right);
    }
}
